doc: gsg cleanups

mbolivar-nordic opened the pull request:
Fix up some error handling for the new zephyr-export extension used in the getting started guide, and do some other cleanups in the GSG.

mbolivar-nordic committed:
doc: gsg: simplifications and fixups

Strip out text that isn't needed to try to minimize the GSG's length.
Fix up some grammar nits.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: gsg: unify 'get zephyr' steps

Various commands are getting put into their own sections when they are
really just steps along the way towards getting zephyr and installing
Python dependencies. Group them together in a section by that name,
moving the west install step there.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -61,116 +50,84 @@ tools. First, make sure your development system OS is updated:
-              make gcc gcc-multilib g++-multilib libsdl2-dev
+            sudo apt install --no-install-recommends cmake ninja-build gperf \
+              python3 python3-setuptools python3-pip python3-wheel git \
+              ccache wget xz-utils file
```
native posix won't work if you don't get the multilibs. Are you sure we want to remove those?

mbolivar-nordic replied:
How about we move those to the toolchain page in that case?

tejlmand replied:
I think the getting started guide should get you all needed tools without looking elsewhere.

And being able to download Zephyr and just test it by building native posix, is one such thing that should be possible out-of-the box.

Just as it makes life easier for Linux users, that Zephyr provides a SDK with toolchains, so you don't need to go to 3rd party toolchain page (which you need to do on windows)

nashif replied:
> How about we move those to the toolchain page in that case?

how is this related to toolchains? Agree with @tejlmand

mbolivar-nordic replied:
Those are the toolchains that are needed to build a particular target (native_posix)... what am I missing?

tejlmand replied:
and so is: `arm`, `arc`, `xtensa`, and all the other toolchains.
Why should they work, and not `native_posix`, when we don't know which toolchain the users will be using with Zephyr ?

carlescufi replied:
Agreed, I suggest leaving the multilib requirements as they were. native_posix is very important to many newcomers to Zephyr.

mbolivar-nordic replied:
I've restored the Ubuntu packages and moved the Python script changes to their own PR, https://github.com/zephyrproject-rtos/zephyr/pull/24248

nashif commented:
for the gsg, this is no cleanup, this is almost a rewrite :)

mbolivar-nordic commented:
> for the gsg, this is no cleanup, this is almost a rewrite :)

That's not totally unfair. In my defense, I will say that there were items in https://github.com/zephyrproject-rtos/zephyr/pull/19123 and https://github.com/zephyrproject-rtos/zephyr/pull/23054 that I had either suggested or intimated should be done "later". Later is now :).

tejlmand commented on scripts/west_commands/export.py:
```diff
@@ -39,38 +42,52 @@ def do_add_parser(self, parser_adder):
+        # The 'share' subdirectory of the top level zephyr repository.
+        share = Path(__file__).parents[2] / 'share'
+
+        if which('ninja') is not None:
```
Just a thought.

We don't need a specific generator for the export, as it doesn't matter which generator is used on that part (build command is never called).
Only reason the build command for the generator used must be installed is the `cmake --build <path>` command.

So, alternatively, we could decide to let Python remove the files, and keeping only those listed.
That way, this test could be removed.

Comments ?

tejlmand commented on scripts/west_commands/export.py:
```diff
@@ -39,38 +42,52 @@ def do_add_parser(self, parser_adder):
+                  capture_output=True)
+    except CalledProcessError:
+        # Do our best to clean up even though CMake failed.
+        log.wrn(f'Failed to make {path} pristine; '
+                'removing known generated files...')
+        for subpath in ['CMakeCache.txt', 'CMakeFiles', 'build.ninja',
+                        'cmake_install.cmake', 'rules.ninja']:
+            remove_if_exists(Path(path) / subpath)
```
We anyway do a cleanup here, so let's not try  to call the pristine.
I can make additional commit with suggestion and you can squash , if you like it.

mbolivar-nordic replied:
I'm happy to take any patches you want to this file! I just ran into a couple of issues and wanted to help.

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -61,116 +50,100 @@ tools. First, make sure your development system OS is updated:
                wget -O - https://apt.kitware.com/keys/kitware-archive-latest.asc 2>/dev/null | sudo apt-key add -
 
-         b) Add the kitware repo corresponding to the Ubuntu 18.04 LTS release:
+         b) Add the Kitware apt repository for your OS release. For Ubuntu
+            18.04:
```
I think `Ubuntu 18.04 LTS` is the right term: https://releases.ubuntu.com/

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -61,116 +50,100 @@ tools. First, make sure your development system OS is updated:
-         doesn't support flashing your application to the board.  As such,
-         we don't recommend using WSL yet.
+         for Zephyr development yet, because it's harder to flash Zephyr
+         applications to boards using WSL.
```
Not supported, or just harder ?
If it is just harder, should we make a page describing, how to do it ?

mbolivar-nordic replied:
A user reported it's possible if you manually rename some of the underlying flash utilities. I haven't figured it out or even tried. I think that's future work and the recommendation should stand; I'm just trying to avoid a potentially inaccurate claim that it "can't" be done since users are claiming it can.

tejlmand replied:
Ok, thanks for the explanation.
Then howabout something like:

```
The Zephyr project does not support flashing of your application using the
built-in Windows Subsystem for Linux (WSL).

Therefore we don't recommend using the `Windows Subsystem for Linux (WSL).
```
that way we only say we are not supporting it, which is different from: It cannot be done.

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -251,41 +213,23 @@ directory using west:
-
-      .. code-block:: bash
+The :ref:`Zephyr CMake package <cmake_pkg>` allows CMake to load boilerplate
+code required for building Zephyr applications.
```
Nothing wrong here, but even if you don't export, then CMake can still load the boilerplate code when using `ZEPHYR_BASE`.
So the main benefit here is 
```suggestion
The :ref:`Zephyr CMake package <cmake_pkg>` allows CMake to automatically load boilerplate
code required for building Zephyr applications.
```

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -273,27 +273,29 @@ to build Zephyr applications.
          .. code-block:: bash
 
-            cd ~
+            cd ~/zephyrproject
```
Personally, I prefer to have my tools / programs outside any repository.

Especially as tools might be up-/downgrade independently of the repo.

Or I could have multiple up-/downstream workspaces, but just a single toolchain.

So I would suggest to still advice users with a different path.

Which of course also means the comment goes for remaining part of this commit.

mbolivar-nordic replied:
I am doing this on purpose for the getting started guide precisely to make the Zephyr environment more self-contained, so we appear to have different preferences in this regard. The question is really what will be easier for more people, I suppose.

tejlmand replied:
could you take a look at:
https://github.com/zephyrproject-rtos/sdk-ng/pull/199
https://github.com/zephyrproject-rtos/zephyr/pull/24143

those will remove the need for the environment variables, and in such way make it easier to get started.
And not polluting the Zephyr workspace with a toolchain folder.

mbolivar-nordic replied:
> could you take a look at:
> [zephyrproject-rtos/sdk-ng#199](https://github.com/zephyrproject-rtos/sdk-ng/pull/199)
> #24143
> 

OK, I like that PR very much. Let's wait until it gets reviewed by @nashif and @galak to decide how to proceed here.

carlescufi replied:
I suggest using either `/opt/` or ~/bin/` instead

tejlmand replied:
> I suggest using either `/opt/` or ~/bin/` instead


I think `~/bin` is a bad option, as the Zephyr SDK is not a single executable file, but several folders with executables, libraries, etc.

Currently, the suggested default paths to search for the SDK are:
```
$HOME/zephyr-sdk*
$HOME/.local/zephyr-sdk*
$HOME/.local/opt/zephyr-sdk*
/opt/zephyr-sdk*
/usr/zephyr-sdk*
/usr/local/zephyr-sdk*
```
so I suggest we pick one among those in the example given in gsg.

Feel free to comment on default paths for SDK in this PR:
https://github.com/zephyrproject-rtos/sdk-ng/pull/199

carlescufi replied:
I meant `~/bin/zephyr-sdk` @tejlmand

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -161,95 +145,118 @@ You'll also install :ref:`west <west>`, Zephyr's multi-purpose tool.
+            west update
 
-      .. code-block:: bat
+      #. Export a :ref:`Zephyr CMake package <cmake_pkg>`. This allows CMake to
+         load boilerplate code required for building Zephyr applications.
```
Changing the changes in later commit in same PR.
So unfriendly of you :wink: 

```suggestion
      #. Export a :ref:`Zephyr CMake package <cmake_pkg>`. This allows CMake to
         automatically load boilerplate code required for building Zephyr applications.
```

mbolivar-nordic replied:
Sorry, I will try to be more friendly in the future. Comments now addressed.

tejlmand reviewed (changes requested):
Thanks for this.

In general I think the cleanup looks good, but still has some suggestions.

mbolivar-nordic commented:
Last force-push is just a rebase. Will address comments after to make it easier for reviewers.

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -86,91 +70,70 @@ We'll also install Zephyr's multi-purpose west tool.
-      `Chocolatey`_, a package manager for Windows.  If you prefer to install
-      dependencies manually, you can also download the required programs from their
-      respective websites and verify they can be found on your PATH.
+      They rely on `Chocolatey`_. If Chocolatey isn't an option, you can
```
```suggestion
      The instructions rely on `Chocolatey`_. If Chocolatey isn't an option, you can
```

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -179,292 +142,276 @@ We'll also install Zephyr's multi-purpose west tool.
-      compatible with your board, a good alternative to try is the
-      :ref:`Hello World sample <hello_world>`.
+   Blinky is compatible with most, but not all, :ref:`boards`. If your board
+   does not meet Blinky's :ref:`blinky-sample-requirements`, :ref:`hello_world`
```
```suggestion
   does not meet Blinky's :ref:`blinky-sample-requirements`, then :ref:`hello_world`
```

carlescufi replied:
This is so that the 2 hyperlinks do not "visually merge"

carlescufi commented on doc/getting_started/index.rst:
```diff
@@ -179,292 +142,276 @@ We'll also install Zephyr's multi-purpose west tool.
-* Learn about :ref:`application` and more details about :ref:`west`.
-* Check out :ref:`beyond-GSG` for information about advanced setup
-  alternatives and issues.
+* Try other :ref:`samples-and-demos`.
```
I suggest also linking to the debugging section here, which is likely one of the things users want to try first.

carlescufi reviewed (changes requested):
I finally had a chance to render this locally and review it. In general this is a great improvement, thank you @mbolivar-nordic. I only have a few suggestions and one section that I'd like changed.

mbolivar-nordic commented on doc/getting_started/index.rst:
```diff
@@ -3,17 +3,11 @@
-- **Get the sources**,
-- **Build, flash, and run** a sample application on your target board.
-
-\*Instructions for other Linux distributions are discussed in the
```
OK, I'll remove it from the first ubuntu tab instead, then. The same link appears twice in consecutive paragraphs, so I was trying to remove one of them. I agree keeping it here makes more sense.

mbolivar-nordic commented:
> I suggest using either `/opt/` or `~/bin/` instead

I will just drop the patch affecting the Zephyr SDK entirely if that is OK, since @tejlmand already has a plan for that. This will minimize user-visible churn during the release cycle.

mbolivar-nordic committed:
doc: blinky: touch up README

The devicetree link is rendering as "Devicetree Guide" instead of
"devicetree". Fix that.

My guess is that most users won't care about the details of the
requirements, and rather just want to know if their board is supported
or not. So move the error you'll see on unsupported boards before the
details about how to add support (which involve a pretty significant
learning curve).

Also mention overlays as a way to get this working.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: hello_world: touch up README

Since this is such a basic sample, it's worth nitpicking the wording a
bit:

- Add a link to the supported boards, since many first time users
  will not know where it is and may start near here.

- Remove the single and multi threaded note. There's no information on
  how to build it in either of these two modes or what the default is,
  so it feels like a distraction for such a basic sample.

- Give a clue about how to build for another board.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
@carlescufi rebased and comments addressed. I added a couple of patches to the main two samples referenced from the guide

tejlmand commented:
> 
> 
> > I suggest using either `/opt/` or `~/bin/` instead
> 
> I will just drop the patch affecting the Zephyr SDK entirely if that is OK, since @tejlmand already has a plan for that. This will minimize user-visible churn during the release cycle.

That is fine with me.
And I think it makes it easier to get this in, and I can rebase #24143 and then adjust the Zephyr SDK accordingly.
Makes it clearer what changes are introduced for which reason.

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -22,18 +18,13 @@ you'll:
 
-Zephyr development depends on an up-to-date host system and common build system
-tools. First, make sure your development system OS is updated:
+Click the operating system you are using.
```
When looking at the rendered doc, I think there should be just a single line explaining why users should do those commands, so maybe add a little:
```suggestion
Click the operating system you are using and follow the guide to ensure it is up-to-date.
```

mbolivar-nordic replied:
I think this is implied by the title "Select and Update OS"

tejlmand replied:
yeah, but when I first read it, I just read the content and my eyes didn't notice the title.
Not sure if i'm the only one :laughing:

mbolivar-nordic replied:
I guess adding more text isn't likely to make eyes glaze less, though. That's exactly why I'm trying to strip this down and remove any text I can -- people generally ignore text and just start copy pasting command lines into their terminals as fast as they can, in my experience.

tejlmand commented on doc/getting_started/index.rst:
```diff
@@ -61,18 +51,13 @@ tools. First, make sure your development system OS is updated:
-applications.
-
-We'll also install Zephyr's multi-purpose west tool.
+Next, you'll install some host dependencies using your package manager.
```
Maybe emphasize the importance ?
```suggestion
Next, you'll install required host dependencies using your package manager.
```

mbolivar-nordic replied:
Well, they are not all required for all use cases, as discussed earlier in the review. Strictly speaking several of these are optional dependencies only needed for some targets or some use cases.

Since I was requested to leave all the optional dependencies I discovered in thisdocument, I have done so, but they're not required ;).

mbolivar-nordic replied:
For reference and the record:

- ccache: this is an optimization
- wget: only needed to pull down the Zephyr SDK, which is optional
- make: optional
- gcc, gcc-multilib, g++-multilib: only needed for some host targets, per review
- dfu-util: only needed for some boards as a runner dependency
- libsdl2-dev: I honestly don't know
- python3-tk: only needed for the guiconfig target, as far as I can tell
- file: only needed by the SDK installer
- xz-utils: needed to unpack the SDK, not required on all toolchains
- device-tree-compiler: optional, just used for additional warnings


You can see that not much is actually required.

tejlmand replied:
Thanks for the list.
Could we add this information somewhere and link to it ?

Personally I like to be able to know why I need to install a certain tool, if it is not strictly required. 
(Note: this can be an independent improvement to the cleanup)

mbolivar-nordic replied:
I'd rather not maintain such a thing and it's likely to go stale. People tend to build additional users on existing dependencies.

tejlmand replied:
point taken.

tejlmand reviewed (approved):
Great cleanup.
Found two small places I think can improve even more, feel free to take them in or ignore them.

mbolivar-nordic commented:
Rebased, hopefully CI is happy now

carlescufi reviewed (approved):
This LGTM now, thank you @mbolivar-nordic, this will be clearer and more accessible for upcoming new users

mbolivar-nordic commented:
Thanks! Rebased, hopefully good to go after CI re-runs.

tejlmand commented:
If we get this in, then I can rebase: https://github.com/zephyrproject-rtos/zephyr/pull/24143
and adjust the getting started changes in that PR accordingly.

mbolivar-nordic commented on doc/getting_started/index.rst:
```diff
@@ -179,294 +143,277 @@ We'll also install Zephyr's multi-purpose west tool.
 
-Get the source code
-*******************
+Install Zephyr
```
> we do not want to make it sound as if Zephyr was an installable application, because it is not.

addressed

mbolivar-nordic commented on doc/getting_started/index.rst:
```diff
@@ -86,91 +71,70 @@ We'll also install Zephyr's multi-purpose west tool.
          .. code-block:: bash
 
-            brew install cmake ninja gperf ccache dfu-util qemu dtc python3
+            brew install cmake ninja gperf python3 ccache
```
That was a leftover from removing optional dependencies. Fixed, thanks.

nashif commented:
probably not fault of this PR, I just tried following the new gsg on new ubuntu 2020.04 and python3-dev is required by some python package we install (psutil)... 
Without the pkg the install fails...

erwango commented on doc/getting_started/index.rst:
```diff
@@ -179,30 +159,23 @@ We'll also install Zephyr's multi-purpose west tool.
-
-         .. code-block:: console
-
-            choco install git python ninja dtc-msys2 gperf
```
@mbolivar-nordic I'm sure this is documented somewhere (as otherwise we'd have new users complaining all day long), but I wonder how windows users are supposed to get dtc if not documented here.

mbolivar-nordic replied:
> @mbolivar-nordic I'm sure this is documented somewhere (as otherwise we'd have new users complaining all day long), but I wonder how windows users are supposed to get dtc if not documented here.

It's not required; 23df708aa36d9b748b54d1c362e7e2e67316593c

mbolivar-nordic replied:
> I'm sure this is documented somewhere 

https://docs.zephyrproject.org/latest/guides/dts/intro.html#scripts-and-tools

erwango replied:
Ho, right! 
I was confused about dtc status, I knew it was still in use (at least on my side).
That's clear now. Txs @mbolivar-nordic
